lib/packetgen/header/http/request.rb

```diff
@@ -0,0 +1,92 @@
+# This file is part of PacketGen
+# See https://github.com/sdaubert/packetgen for more informations
+# Copyright (C) 2016 Sylvain Daubert <sylvain.daubert@laposte.net>
+# This program is published under MIT license.
+
+module PacketGen
+  module Header
+    module HTTP
+      # An HTTP/1.1 Request packet consits of:
+      # * the http method ({Types::String}).
+      # * the path ({Types::String}).
+      # * the version ({Types::String}).
+      # * associated http headers ({Types::String}).
+      #
+      # == Create a HTTP Request header
+      #   # standalone
+      #   http_rqst = PacketGen::Header::HTTP::Request.new
+      #   # in a packet
+      #   pkt = PacketGen.gen("IP").add("TCP").add("HTTP::Request")
+      #   # access to HTTP Request header
+      #   pkt.http_request # => PacketGen::Header::HTTP::Request
+      #
+      # == HTTP Request attributes
+      #	  http_rqst.version = "HTTP/1.1"
+      #	  http_rqst.method = "GET"
+      #	  http_rqst.path = "/meow.html"
+      #   http_rqst.headers = "Host: tcpdump.org"     # string or
+      #	  http_rqst.headers = { host: "tcpdump.org" } # even a hash
+      #
+      # @author Kent 'picat' Gruber
+      class Request < Base
+        # @!attribute method
+        #   @return [Types::String]
+        define_field :method,  Types::String
+        # @!attribute path 
+        #   @return [Types::String]
+        define_field :path,    Types::String
+        # @!attribute version 
+        #   @return [Types::String]
+        define_field :version, Types::String, default: "HTTP/1.1"
+        # @!attribute headers
+        #   associated http/1.1 headers
+        #   @return [Types::HTTPHeaders]
+        define_field :headers, Types::HTTPHeaders 
+
+        # @param [Hash] options
+        # @option options [String] :method
+        # @option options [String] :path
+        # @option options [String] :version
+        # @option options [Hash]   :headers
+        def initialize(options={})
+          super(options)
+          self.headers ||= options[:headers]
+        end
+
+        # Give it a pretty name, the only way I know how.
+        # @return [String]
+        def protocol_name
+          "HTTP_Request"
+        end
+
+        # Read in the HTTP portion of the packet, and parse it. 
+        # @return [PacketGen::HTTP::Request]
+        def read(str)
+          # prepare data to parse
+          str = str.split("\n").map(&:strip).reject(&:empty?)
+          first_line = str.shift.split
```

Comment:
Use another name than `str`, as this is no more a string but an array.
You may use `String#lines` instead of `#split`